research/clustering/information_abstraction.py

```diff
@@ -0,0 +1,439 @@
+        That logic is used within the literature
+
+        :param start_combos: starting combination of cards (beginning with hole cards)
+        :param publics: np.array of public combinations being added
+        :return: Combinations of private information (hole cards) and public information (board)
+        """
+        our_cards = []
+        for combos in start_combos:
+            for public_combo in publics:
+                # TODO: need a way to create these combos with better performance?
+                if not np.any(np.isin(combos, public_combo)):
+                    our_cards.append(np.concatenate((combos, public_combo)))
+        return np.array(our_cards)
+
+
+class InfoBucketMaker(InfoSets):
+    """
+    Stores info buckets for each street when called
+    # TODO: create key to access these from a dictionary, store more efficiently somehow
+    # TODO: change cluster to num_clusters=200 for full deck
+    """
+    def __init__(self, num_cards):
+        super().__init__(num_cards)
+
+        self._river_ehs = self.get_river_ehs(num_print=1000)
+        self._river_centroids, self._river_clusters = self.cluster(
+            num_clusters=15, X=self._river_ehs
+        )
+        self._turn_ehs_distributions = self.get_turn_ehs_distributions(num_print=100)
+        self._turn_centroids, self._turn_clusters = self.cluster(
+            num_clusters=15, X=self._turn_ehs_distributions
+        )
+        self._flop_potential_aware_distributions = self.get_flop_potential_aware_distributions(
+            num_print=100
+        )
+        self._flop_centroids, self._flop_clusters = self.cluster(
+            num_clusters=15, X=self._flop_potential_aware_distributions
+        )
+
+    def __call__(self):
+        # TODO: switch to log
+        self.dump_data(location="data/information_abstraction.pkl")
+        self.print_cluster_example(X=self._river_ehs, clusters=self._river_clusters,
+                                   cluster_name="Expected Hand Strength on River", cluster_id=4)
+        self.print_cluster_example(X=self._turn_ehs_distributions, clusters=self._turn_clusters,
+                                   cluster_name="Expected Hand Strength Distribution on Turn", cluster_id=4)
+        self.print_cluster_example(X=self._turn_ehs_distributions, clusters=self._turn_clusters,
+                                   cluster_name="Potential Aware Distribution on Flop", cluster_id=4)
+        self.plot_river_clusters()
+
+    @staticmethod
+    def simulate_get_ehs(game: GameUtility, num_simulations: int = 25) -> List[float]:
+        """
+        # TODO: probably want to increase simulations..
+        :param game: GameState for help with determining winner and sampling opponent hand
+        :param num_simulations: how many simulations you want to do
+        :return: [win_rate, loss_rate, tie_rate]
+        """
+        ehs = [0] * 3
+        for _ in range(num_simulations):
+
+            idx = game.get_winner()
```

Comment:
Is there some code that is missing here? Here we repeatedly call `game.get_winner()` for `num_simulations` but internally in the `GameUtility` class the `self.opp_hand` and `self.our_hand` don't appear to change in value? 

It would almost be good to add a function to `GameUtility` that would allow us for a given hand to iterate over all possible opponent hands.

E.g:
```python
for ehs_index in game.get_all_results():
    ehs[ehs_index] += 1 / game.n_simulations
```

Reply:
Actually scratch that, I can see you have defined the `self.opp_hand` to be a random sampling via a property decorated method. Nice!

Still might be interesting if we could deterministically work through every possible hand. Not sure how brutal combinatorically that might be though1

Reply:
It's a good thought, but I think we'd get sufficiently close with simulations. On the flop that would be ~5.56*10^13, if considering every possible opponent hand, then turn, then river. Although it would be much smaller if we moved to lossless

Reply:
Probably I'll look into lossless early next week or perhaps try to work out the trade off of using lossless versus seeing at what # of simulations we start getting diminishing returns on more samples/running deterministically, all that jazz